src/main/java/org/folio/service/inventory/InventoryInstanceManager.java

```diff
@@ -347,17 +343,46 @@ private Future<String> createInstance(JsonObject instanceRecJson, RequestContext
   public Future<CompositePoLine> openOrderHandleInstance(CompositePoLine compPOL, boolean isInstanceMatchingDisabled, RequestContext requestContext) {
     return consortiumConfigurationService.getConsortiumConfiguration(requestContext)
       .compose(consortiumConfiguration -> {
-        if (compPOL.getInstanceId() != null) {
-          return consortiumConfiguration.isEmpty() ? Future.succeededFuture(compPOL.getInstanceId())
-            : shareInstanceAmongTenantsIfNeeded(compPOL.getInstanceId(), consortiumConfiguration.get(), compPOL.getLocations(), requestContext);
-        } else {
-          return getInstanceRecord(compPOL, isInstanceMatchingDisabled, requestContext)
-            .compose(instanceId -> consortiumConfiguration.isEmpty() ? Future.succeededFuture(instanceId)
-              : shareInstanceAmongTenantsIfNeeded(instanceId, consortiumConfiguration.get(), compPOL.getLocations(), requestContext));
-        }
+        Future<String> instanceIdFuture = compPOL.getInstanceId() != null ?
+          Future.succeededFuture(compPOL.getInstanceId()) :
+          getInstanceRecord(compPOL, isInstanceMatchingDisabled, requestContext);
+        consortiumConfiguration.ifPresent(
+          configuration -> instanceIdFuture
+            .compose(instanceId -> shareInstanceAmongTenantsIfNeeded(instanceId, configuration, compPOL.getLocations(), requestContext))
+        );
+        return instanceIdFuture;
       }).map(compPOL::withInstanceId);
   }
 
+  private Future<String> shareInstanceAmongTenantsIfNeeded(String instanceId, ConsortiumConfiguration consortiumConfiguration,
+                                                           List<Location> locations, RequestContext requestContext) {
+    return findTenantsWithUnsharedInstance(instanceId, locations, requestContext)
+      .map(tenantIds -> tenantIds.stream().map(targetTenantId -> sharingInstanceService.createShadowInstance(instanceId,
+        targetTenantId, consortiumConfiguration, requestContext)).toList())
+      .compose(HelperUtils::collectResultsOnSuccess)
+      .map(sharingInstances -> instanceId);
+  }
+
+  /*
+  Returns list of tenant ids for which instance wasn't found
+   */
+  private Future<List<String>> findTenantsWithUnsharedInstance(String instanceId, List<Location> locations, RequestContext requestContext) {
+    List<Future<Optional<String>>> tenantIdFutures = locations.stream()
+      .map(Location::getTenantId)
+      .distinct()
+      .filter(Objects::nonNull)
+      .map(tenantId -> RequestContextUtil.createContextWithNewTenantId(requestContext, tenantId))
+      .map(
+        clonedRequestContext -> getInstanceById(instanceId, false, clonedRequestContext)
+        .map(instance -> Optional.<String>empty())
+        .recover(throwable -> throwable instanceof HttpException httpException && httpException.getCode() == 404 ?
+          Future.succeededFuture(Optional.of(TenantTool.tenantId(clonedRequestContext.getHeaders()))) : Future.failedFuture(throwable))
```

Comment:
ternary operationg it is usuall better for small line of code. Suggest to move if statements

Reply:
```
if (throwable instanceof HttpException httpException && httpException.getCode() == 404) {
  return Future.succeededFuture(Optional.of(TenantTool.tenantId(clonedRequestContext.getHeaders())));
}
return Future.failedFuture(throwable))
```

Reply:
Old code as well. Actually I agree that many parts of orders codebase can be simplified.

Reply:
it is really appreciated to simply the code, of course if time allows to do so

Reply:
done